### PR TITLE
New and improved lock LED callbacks

### PR DESCRIPTION
**Pull request by fauxpark**
## Description

This uses a struct to store the LED state, which is much less verbose than doing bitwise operations on a `uint8_t`, even with the `IS_LED_ON()` macros. Since there is no overriding in C, it necessarily required a new set of callbacks, `bool led_update_*(led_t)`. As you can see, it returns a boolean instead of void, so that keymaps can overrule the keyboard-level implementation. I've provided an example for the Wasdat, and updated the documentation and templates.

I don't feel this is a breaking change, but it may be a good idea to tag it so it gets more visibility.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: docs/custom_quantum_functions.md [hunk 1]
```diff
@@ -90,68 +90,68 @@ keyrecord_t record {
...
 
-### Example `led_set_user()` Implementation
+!> `host_keyboard_led_state()` may already reflect a new value before `led_update_user()` is called.
+
+### Example `led_update_user()` Implementation
 
 ```c
-void led_set_user(uint8_t usb_led) {
-    if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
+bool led_update_user(led_t led_state) {
```
Comment (yanfali) [inline docs/custom_quantum_functions.md]: does this go in keyboard.c or keymap.c?

**Comment by yanfali** [comment]
I feel like this is a breaking change, as you're deprecating an interface we've been promoting. What does older code do?

**Comment by drashna** [comment]
It looks like the old code will still work.  However, it is deprecating it in favor of a better interface.

I'm not sure that breaking changes here applies, since the old functions are not removed, nor are they altered.

**Comment by yanfali** [comment]
Can we use something like doxygen comments to make them as deprecated, and also write something up in the docs?

File: docs/custom_quantum_functions.md (hunk 1 above)
Comment (fauxpark) [reply docs/custom_quantum_functions.md]: It's under the "Function Documentation" heading - just following the format of this page.

**Comment by fauxpark** [comment]
@yanfali how's this?

Might need some wordsmithing...

File: keyboards/maartenwut/wasdat/wasdat.c [hunk 2]
```diff
@@ -33,26 +33,12 @@ void led_init_ports(void) {
...
-    } else {
-        writePinHigh(B1);
-    }
-
-    if (IS_LED_ON(usb_led, USB_LED_NUM_LOCK)) {
-        writePinLow(B2);
-    } else {
-        writePinHigh(B2);
+bool led_update_kb(led_t led_state) {
+    if(led_update_user(led_state)) {
```
Comment (drashna) [inline keyboards/maartenwut/wasdat/wasdat.c]: Oh my, YES!

File: docs/custom_quantum_functions.md [hunk 3]
```diff
@@ -90,68 +90,75 @@ keyrecord_t record {
...
     } else {
         writePinHigh(B3);
     }
-    if (IS_LED_ON(usb_led, USB_LED_KANA)) {
+    if (led_state.kana) {
         writePinLow(B4);
     } else {
         writePinHigh(B4);
     }
+    return true;
```
Comment (drashna) [inline docs/custom_quantum_functions.md]: Should mention somewhere, that returning true allows for normal processing, and false replaces the effect.  Depending on the board.

File: docs/custom_quantum_functions.md [hunk 4]
```diff
@@ -98,18 +98,25 @@ QMK provides methods to read 5 of the LEDs defined in the HID spec:
...
+There are two ways to get the lock LED state:
 
 * by implementing `led_update_kb()` or `_user()`; or
 * by calling `host_keyboard_led_state()`
 
+!> `host_keyboard_led_state()` may already reflect a new value before `led_update_user()` is called.
+
+Two more deprecated functions exist that provide the LED state as a `uint8_t`:
+
+* `led_set_kb()` and `_user()`
```
Comment (yanfali) [inline docs/custom_quantum_functions.md]: let's put the correct c signature in here. I would also like to mark the code. quantum.c (L1069) and quantum.h. I would add a TODO tag with a date to remove these calls or indicate they are now deprecated. People rarely read docs but they look at code.

Comment (fauxpark) [reply docs/custom_quantum_functions.md]: Done. I already added `\deprecated` doc comments.

File: docs/custom_quantum_functions.md (hunk 3 above)
Comment (fauxpark) [reply docs/custom_quantum_functions.md]: I was thinking that's what the sentence "You must return either `true` or `false` from this function, depending on whether you want to override the keyboard-level implementation." conveyed, but maybe it's not clear enough? I'm having trouble getting this whole section to flow right TBH.

**Pull request merged**